torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -449,6 +479,92 @@ void ProcessGroupAgent::listenLoop() {
   }
 }
 
+void ProcessGroupAgent::pollTimedOutRPCs() {
+  while (!shutdown_.load()) {
+    std::chrono::milliseconds sleepTime;
+    std::unique_lock<std::mutex> lock{futureMutex_};
+    // Estimate amount of time the first future will time out in, and sleep
+    // for that long.
+    // if there are no futures or the RPC timeout is set to 0 (meaning no
+    // timeout), then sleep for a set "infinity" time.
+    if (futureTimeouts_.empty() || rpcTimeout_.load().count() == 0) {
+      sleepTime = INFINITY_SLEEP_INTERVAL;
+    } else {
+      const auto remainingTime = getRPCRemainingTime(
+          futureTimeouts_.begin()->first, rpcTimeout_.load());
+      sleepTime = std::max(remainingTime, std::chrono::milliseconds(0));
+    }
+    futureTimeoutCV_.wait_for(lock, sleepTime);
+
+    if (rpcTimeout_.load().count() == 0) {
```

Comment:
The checks for `rpcTimeout` and `shutdown` doesn't need the lock, but it is needed in `processTimedOutFutures`. Should we do a `lock.unlock()` here and re-acquire it before calling `processTimedOutFutures`? Not sure if the performance difference would be meaningful

Reply:
Don't worry about it since those checks are pretty cheap.